File: Python3/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):    
    #선물 횟수 및 선물 지수 저장
    giftCnt = [[0] * len(friends) for i in range(len(friends))] #선물 횟수
    giftIndex = [0] * len(friends) #선물 지수
    for gift in gifts:
        a, b = gift.split()
        giftCnt[friends.index(a)][friends.index(b)] += 1
        giftIndex[friends.index(a)] += 1
        giftIndex[friends.index(b)] -= 1
    
    #선물 받을 횟수 count
    willGift = [0] * len(friends) #선물 받을 횟수
    for i, f in enumerate(giftCnt):  #각 친구들별 선물 준 횟수 돌리기
        for j, gf in enumerate(f): #선물 받는 친구 구하기
            if gf > giftCnt[j][i] or gf == giftCnt[j][i] and giftIndex[i] > giftIndex[j]:
                willGift[i]+=1
        
    return max(willGift)

File: C++/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <sstream>

using namespace std;

void splitName(string gift, vector<string> *names) {
    istringstream ss(gift);
    int i = 0;
    string buffer;
    while(getline(ss, buffer, ' ')){
        (*names)[i++] = buffer;
    }
}

int solution(vector<string> friends, vector<string> gifts) {
    int N = friends.size();
    //각 이름별 인덱스 설정
    map<string ,int> nameMap;
    for(int i = 0; i < N; i++) {
        nameMap[friends[i]] = i;
    }
    
    // 선물 횟수 및 지수 초기화
    vector<vector<int>> giftCnt(N, vector<int>(N));
    vector<int> giftIndex(N);
    vector<string> names(2);
    int a, b;
    for(string gift : gifts) {
        splitName(gift, &names);
        a = nameMap[names[0]];
        b = nameMap[names[1]];
        giftCnt[a][b]++;
        giftIndex[a]++;
        giftIndex[b]--;
    }
    
    //선물 비교
    vector<int> compare(N);
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < N; j++) {
            if(giftCnt[i][j] > giftCnt[j][i]
               || giftCnt[i][j] == giftCnt[j][i] && giftIndex[i] > giftIndex[j]) {
                compare[i]++;
            }
        }
    }
    
    sort(compare.begin(), compare.end());
    return compare[N - 1];
}